erts/emulator/beam/jit/x86/instr_arith.cpp

```diff
@@ -581,19 +639,82 @@ void BeamGlobalAssembler::emit_int_div_rem_body_shared() {
 void BeamModuleAssembler::emit_div_rem(const ArgVal &Fail,
                                        const ArgVal &LHS,
                                        const ArgVal &RHS,
-                                       const ErtsCodeMFA *error_mfa) {
-    mov_arg(ARG4, RHS); /* Done first as mov_arg may clobber ARG1 */
-    mov_arg(ARG1, LHS);
+                                       const ErtsCodeMFA *error_mfa,
+                                       bool need_div,
+                                       bool need_rem) {
+    Label generic_div = a.newLabel(), next = a.newLabel();
+    bool need_generic = true;
+    Uint rhs = RHS.getValue();
+    Sint divisor = 0;
+
+    if (RHS.isImmed() && is_small(rhs)) {
+        divisor = signed_val(rhs);
+    }
 
-    /* TODO: Specialize division with immediates, either here or in the
-     * compiler. */
-    if (Fail.getValue() != 0) {
-        safe_fragment_call(ga->get_int_div_rem_guard_shared());
-        a.je(resolve_beam_label(Fail));
-    } else {
-        a.mov(ARG5, imm(error_mfa));
-        safe_fragment_call(ga->get_int_div_rem_body_shared());
+    if (divisor != (Sint)0 && divisor != (Sint)-1) {
+        /* There is no possibility of overflow. */
+        a.mov(ARG6, imm(divisor));
+        mov_arg(x86::rax, LHS);
+        if (always_small(LHS)) {
+            comment("skipped test for small dividend since it is always small");
+            need_generic = false;
+        } else if (always_one_of(LHS, BEAM_TYPE_FLOAT | BEAM_TYPE_INTEGER)) {
+            comment("simplified test for small dividend since it is an "
+                    "integer");
+            a.mov(ARG2d, x86::eax);
+            a.test(ARG2d, imm(TAG_PRIMARY_LIST));
+            a.short_().je(generic_div);
+        } else {
+            comment("testing for a small dividend");
+            a.mov(ARG2d, x86::eax);
+            a.and_(ARG2d, imm(_TAG_IMMED1_MASK));
+            a.cmp(ARG2d, imm(_TAG_IMMED1_SMALL));
+            a.short_().jne(generic_div);
+        }
+
+        /* Sign-extend and divide. The result is implicitly placed in
+         * RAX and the remainder in RDX (ARG3). */
+        comment("divide with inlined code");
+        a.sar(x86::rax, imm(_TAG_IMMED1_SIZE));
+        a.cqo();
+        a.idiv(ARG6);
+
+        if (need_div) {
+            a.sal(x86::rax, imm(_TAG_IMMED1_SIZE));
+        }
+
+        if (need_rem) {
+            a.sal(x86::rdx, imm(_TAG_IMMED1_SIZE));
+        }
+
+        if (need_div) {
+            a.or_(x86::rax, imm(_TAG_IMMED1_SMALL));
+        }
+
+        if (need_rem) {
+            a.or_(x86::rdx, imm(_TAG_IMMED1_SMALL));
+        }
```

Comment:
Why 4 ifs and not just 2?

Reply:
I have kept the original order of the instructions, thinking that perhaps it could enable more instruction-level parallelism. Not sure whether it makes any difference, or whether the CPU itself can do the same reordering.

Reply:
[GCC and Clang](https://godbolt.org/z/rT9hhW4a4) seem to think that it does matter, so let us leave it like this.